File: core/src/main/java/javabot/dao/NickServDao.java
package javabot.dao;

import javabot.model.NickServInfo;
import javabot.model.criteria.NickServInfoCriteria;
import javabot.model.criteria.NickServInfoCriteria.NickServInfoUpdater;
import org.pircbotx.User;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

import static javabot.model.NickServInfo.NSERV_DATE_FORMAT;

public class NickServDao extends BaseDao<NickServInfo> {
    protected NickServDao() {
        super(NickServInfo.class);
    }

    public void clear() {
        ds.delete(getQuery());
    }

    public void process(final List<String> list) {
        NickServInfo info = new NickServInfo();
        String[] split = list.get(0).split(" ");
        info.setNick(split[2].toLowerCase());
        info.setAccount(split[4].substring(0, split[4].indexOf(')')).toLowerCase());
        list.subList(1, list.size()).stream().filter(line -> line.contains(":")).forEach(line -> {
            int i = line.indexOf(':');
            String key = line.substring(0, i).trim();
            String value = line.substring(i + 1).trim();
            if (key.equalsIgnoreCase("Registered")) {
                info.setRegistered(extractDate(value));
            } else if (key.equalsIgnoreCase("User Reg.")) {
                info.setUserRegistered(extractDate(value));
            } else if (key.equalsIgnoreCase("Last seen")) {
                info.setLastSeen(extractDate(value));
            } else if (key.equalsIgnoreCase("Last addr")) {
                info.setLastAddress(value);
            } else {
                info.extra(key.replace(".", ""), value);
            }
        });

        NickServInfo nickServInfo = find(info.getAccount());

        if (nickServInfo == null) {
            nickServInfo = find(info.getNick());
        }

        if (nickServInfo != null) {
            nickServInfo.setNick(info.getNick());
            nickServInfo.setLastSeen(info.getLastSeen());
            nickServInfo.setLastAddress(info.getLastAddress());
            save(nickServInfo);
        } else {
            save(info);
        }
    }

    private LocalDateTime extractDate(final String line) {
        if (line.endsWith("now")) {
            return LocalDateTime.now();
        } else {
            String dateString = line.contains("(") ? line.substring(0, line.indexOf(" (")) : line;
            return LocalDateTime.parse(dateString, DateTimeFormatter.ofPattern(NSERV_DATE_FORMAT));
        }
    }

    public NickServInfo find(final String name) {
        NickServInfoCriteria criteria = new NickServInfoCriteria(ds);
        criteria.or(
                       criteria.nick(name.toLowerCase()),
                       criteria.account(name.toLowerCase())
                   );
        return criteria.query().get();
    }

    public NickServInfo updateNick(final String oldNick, final String newNick) {
        NickServInfoCriteria criteria = new NickServInfoCriteria(ds);
        criteria.nick(oldNick);
        NickServInfoUpdater updater = criteria.getUpdater();
        updater.nick(newNick);
        updater.update();
        criteria = new NickServInfoCriteria(ds);
        criteria.nick(newNick);
        return criteria.query().get();
    }

    public void unregister(final User user) {
        NickServInfoCriteria criteria = new NickServInfoCriteria(ds);
        criteria.nick(user.getNick());
        criteria.delete();
    }
}

File: src/main/kotlin/javabot/dao/NickServDao.kt
package javabot.dao

import com.google.inject.Inject
import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.UpdateOptions
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.or
import dev.morphia.query.updates.UpdateOperators.set
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale
import javabot.model.JavabotUser
import javabot.model.NickServInfo

open class NickServDao @Inject constructor(ds: Datastore) :
    BaseDao<NickServInfo>(ds, NickServInfo::class.java) {
    fun clear() = getQuery().delete(DeleteOptions().multi(true))

    fun process(list: List<String>) {
        val info = NickServInfo()
        val split = list[0].split(" ")
        info.nick = split[2].lowercase(Locale.getDefault())
        info.account = split[4].substring(0, split[4].indexOf(')')).lowercase(Locale.getDefault())
        list
            .subList(1, list.size)
            .filter { line -> line.contains(":") }
            .forEach { line ->
                val (key, value) = line.split(':')
                when (key.uppercase(Locale.getDefault())) {
                    "REGISTERED" -> info.registered = extractDate(value)
                    "USER REG." -> info.userRegistered = extractDate(value)
                    "LAST SEEN" -> info.lastSeen = extractDate(value)
                    "LAST ADDR" -> info.lastAddress = value
                    else -> info.extra(key.replace(".", ""), value)
                }
            }
        var nickServInfo: NickServInfo? = find(info.account)

        if (nickServInfo == null) {
            nickServInfo = find(info.nick)
        }

        if (nickServInfo != null) {
            nickServInfo.nick = info.nick
            nickServInfo.lastSeen = info.lastSeen
            nickServInfo.lastAddress = info.lastAddress
            save(nickServInfo)
        } else {
            save(info)
        }
    }

    private fun extractDate(line: String): LocalDateTime {
        return if (!line.endsWith("now")) {
            val dateString = if (line.contains("(")) line.substring(0, line.indexOf(" (")) else line
            LocalDateTime.parse(
                dateString,
                DateTimeFormatter.ofPattern(NickServInfo.NSERV_DATE_FORMAT),
            )
        } else {
            LocalDateTime.now()
        }
    }

    open fun find(name: String): NickServInfo? {
        return ds.find(NickServInfo::class.java)
            .filter(
                or(
                    eq("nick", name.lowercase(Locale.getDefault())),
                    eq("account", name.lowercase(Locale.getDefault())),
                )
            )
            .first()
    }

    fun updateNick(oldNick: String, newNick: String): NickServInfo? {
        ds.find(NickServInfo::class.java)
            .filter(eq("nick", oldNick))
            .update(set("nick", newNick))
            .execute(UpdateOptions().multi(false))

        return ds.find(NickServInfo::class.java).filter(eq("nick", newNick)).first()
    }

    fun unregister(user: JavabotUser) {
        ds.find(NickServInfo::class.java).filter(eq("nick", user.nick)).delete()
    }
}
